docs/proposals/20220209-machinepool-machines.md

```diff
@@ -0,0 +1,196 @@
+
+As another example, currently each MachinePool infrastructure provider must implement their own node drain functionality. When MachinePool Machines become available in the CAPI API, all providers can use its existing node drain implementation.
+
+### Goals
+
+- Enable CAPI MachinePool to create Machines to represent infrastructure provider MachinePool Machines
+- Ensure that Machines in MachinePools can participate in the same behaviors as MachineDeployment Machines (node drain, machine health checks, retired node deletion)
+- Enhance `clusterctl describe` to display MachinePool Machines
+- Allow CAPI MachinePool to include deployment strategies mirroring those of MachineDeployments (future work)
+
+### Non-Goals/Future Work
+
+- Require infrastructure providers to implement MachinePool Machines. This is intended to be an "opt-in" feature, since there are existing MachinePool implementations that cannot support it. For example, `AWSManagedMachinePool` does not allow for deleting individual instances.
+- Implement Cluster Autoscaler MachinePool integration (a separate task, although we have a working POC)
+- Move the MachinePools implementation into a `exp/machinepools` package. The current code location makes its separation from `exp/addons` unclear and this would be an improvement.
+- Improve bootstrap token handling for MachinePools. It would be preferable to have a single, short-lived token for each Machine instead of the current shared, long-lived token.
+- Remove the `ProviderIDList` field from MachinePools. While the initial proposal using an `InfrastructureRefList` made that field arguably unnecessary, it is still used by cluster-autoscaler and perhaps in other integrations.
+## Proposal
+
+To create MachinePool Machines, a MachinePool in CAPI needs information about the instances or replicas associated with the provider's implementation of the MachinePool. This information is attached to the provider's MachinePool infrastructure resource as new, optional status fields `InfrastructureSelector`. This field may be populated by the infrastructure provider.
+
+```golang
+// FooMachinePoolStatus defines the observed state of FooMachinePool.
+type FooMachinePoolStatus struct {
+	// InfrastructureSelector is a label query over FooMachinePoolMachines.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors
+	// +optional
+	InfrastructureSelector metav1.LabelSelector `json:"infrastructureSelector,omitempty"`
+}
+```
+
+If this field is populated, CAPI's MachinePool controller will query for the provider-specific infrastructure resources. That query uses the Selector field, the API version of the \<Provider\>MachinePool, and the Kind of the \<Provider\>MachinePool with "Machine" appended to it. For example, if the \<Provider\>MachinePool is of Kind "FooMachinePool", the query will be of Kind "FooMachinePoolMachine".
```

Comment:
may be elaborate a bit here about contract and expectations? e.g FooMachinePoolMachine will need to satisfy the Machine controller expectations providerID, addresses readiness singal...

Reply:
Just to confirm my understanding, are we introducing `AzureMachinePoolMachine` kind? would it be simpler and less confusing for the users to use the `AzureMachine` instead? (I guess this requires the `AzureMachine` to be modified to behave accordingly to the fact that it is owned by a MachinePool...)

Reply:
> are we introducing AzureMachinePoolMachine kind?

Yes, `AzureMachinePoolMachine` or `DockerMachinePoolMachine`, for example. Azure actually already has such a thing, even before this proposal, and we propose to add such a resource to Docker. It is possible that an existing provider resource like `AzureMachine` could be modified to support both, but we didn't do that in Azure at least.

I need to check on AWS to see if this simple name generation works there as well. IIRC they may have reused AWSMachine.

Reply:
As we discussed in last Friday's feedback meeting, we don't want to specify that providers must create a new resource type to implement this proposal. It should be possible to reuse an existing InfraMachine resource type here, so I've added an `InfrastructureMachineKind` field and updated the proposal.